### PR TITLE
Drift king1998 a3

### PR DESCRIPTION
**Pull request by DriftKing1998**
Finished and ready for Feedback! ;)

### CODE REVIEW

**Comment by a01508252** [comment]
I downloaded both files (functions.py and DriftKing1998_A3.py) and ran DriftKing1998_A3.py. The results matched those in the test files.

In the functions.py file, there are six functions: parse_file, fill_matrix_hard, fill_matrix_easy, backtrace, backtrace_hard, and fill_matrix_easy. I think it would be helpful to add a docstring that clarifies the functions' purpose and expected inputs since there are many functions.

The fill_matrix_hard and fill_matrix_easy functions fill a matrix with values. Fill_matrix_hard allows diagonal movements. Both functions have a similar structure and share a significant amount of code. To reduce redundancy, I suggest combining these functions into a single function. In your code, diagonal movements are always considered if they are present. Instead, try using a flag to indicate whether diagonal movements are allowed or not, as it was one of the requirements.

The backtrace and backtrace_hard functions determine the optimal path through a matrix. Like the fill_matrix functions, these functions share a lot of code. I recommend combining these functions, as I suggested for fill_matrix, to improve maintainability and readability.

In the DriftKing1998_A3.py file, the main script is responsible for parsing command-line arguments, reading an input file, and using the functions defined in functions.py to process the data. The script is well-structured, and the use of argparse to handle command-line arguments is appropriate.

In summary, the code is generally well-structured and performs its intended tasks. However, I see some opportunities for improvement, such as combining similar functions, adding docstrings and comments, and considering diagonals only if a flag is set.

**Comment by DriftKing1998** [comment]
Yeah you are right about the redundant code. I'll try to fix it tomorrow. 

The reason my function does not need the -d flag is since I did not read the instructions to well and therefore constructed my parser in a way that it can read all inputs. That was an error as it took more time to come up with it. I will be sure to read the instructions more attentive next time.

Thanks for the review
